src/libsyntax/ext/tt/macro_check.rs

```diff
@@ -0,0 +1,342 @@
+    /// The stack of Kleene operators (outermost first) where we entered the macro definition.
+    ops: &'a [KleeneToken],
+}
+
+/// Checks that meta-variables are bound and used correctly for each left and right hand-side pairs
+/// of a macro definition. The span is used when no span information is available. In presence of
+/// nested macro definitions (a macro that defines a macro), this check may have false positives and
+/// false negatives.
+pub fn check_meta_variables(
+    sess: &ParseSess,
+    span: Span,
+    lhses: &[TokenTree],
+    rhses: &[TokenTree],
+) -> bool {
+    if lhses.len() != rhses.len() {
+        sess.span_diagnostic.span_bug(span, "length mismatch between LHSes and RHSes")
+    }
+    let mut valid = true;
+    for (i, lhs) in lhses.iter().enumerate() {
+        let rhs = &rhses[i];
+        let mut binders = Binders::default();
+        check_binders(sess, lhs, &[], &mut binders, &[], &mut valid);
+        check_occurrences(sess, rhs, &[], &binders, &[], &mut valid);
+    }
+    valid
+}
+
+/// Checks that `lhs` does not contain duplicate binders (e.g. `$a:expr, $a:expr` since it would be
+/// ambiguous which `$a` to use if we ever needed to) according to `binders` and `macros`. Extends
+/// `binders` with new binders according to `ops`, the stack of Kleene operators. Sets `valid` to
+/// false if an error is encountered.
+fn check_binders(
+    sess: &ParseSess,
+    lhs: &TokenTree,
+    macros: &[MacroState<'_>],
+    binders: &mut Binders,
+    ops: &[KleeneToken],
+    valid: &mut bool,
+) {
+    match *lhs {
+        TokenTree::Token(..) => {}
+        TokenTree::MetaVar(span, name) => {
+            if macros.is_empty() {
+                sess.span_diagnostic.span_bug(span, "unexpected MetaVarDecl in rhs");
+            }
+            if get_binder_info(macros, binders, name).is_none() {
+                binders.insert(name, BinderInfo { span, ops: ops.to_owned() });
+            } else {
+                check_occurrences(sess, lhs, macros, binders, ops, valid);
+            }
+        }
+        TokenTree::MetaVarDecl(span, name, _kind) => {
+            if let Some(prev_info) = get_binder_info(macros, binders, name) {
+                // TODO: If macros.len() > 0, notify that it may be a false positive.
+                sess.span_diagnostic
+                    .struct_span_err(span, "duplicate matcher binding")
+                    .span_note(prev_info.span, "previous declaration was here")
+                    .emit();
+                *valid = false;
+            } else {
+                binders.insert(name, BinderInfo { span, ops: ops.to_owned() });
+            }
+        }
+        TokenTree::Delimited(_, ref del) => {
+            for tt in &del.tts {
+                check_binders(sess, tt, macros, binders, ops, valid);
+            }
+        }
+        TokenTree::Sequence(_, ref seq) => {
+            let mut ops = ops.to_owned();
+            ops.push(seq.op.clone());
```

Comment:
I agree I don't like too much cloning all those slices to vectors. In a functional programming language we would use lists and it would be fine because all prefixes are constant. I added a FIXME to test that.